### PR TITLE
Fix windows race condition when writing image with duplicate layers

### PR DESCRIPTION
**Pull request by dgannon991**
This is a potential solution for #1920, any feedback would be greatly appreciated.

### CODE REVIEW

**Comment by slonopotamus** [comment]
I'd say a more robust solution is "try to rename, if it fails, but destination exists, just remove the source and pretend you're fine". This would also cover the case when multiple `crane` processes stomp on the same file.

**Comment by slonopotamus** [comment]
I had a _very_ similar issue in [git-lfs](https://github.com/git-lfs/git-lfs) in the past. Here's the analysis: https://github.com/git-lfs/git-lfs/pull/3880#issue-512345192 and here's the final fix that solved the issue: https://github.com/git-lfs/git-lfs/commit/662a6248198594a59c5a652b1ab543d8948c302a

**Comment by dgannon991** [comment]
I'll give that a go, thank you for the feedback.

**Comment by dgannon991** [comment]
@slonopotamus I've written this up and given it a test, and I can confirm it works:
```go
	err = os.Rename(w.Name(), renamePath)
	var badInputErr *os.LinkError
	if err != nil && errors.As(err, &badInputErr) && badInputErr.Err == syscall.ERROR_ACCESS_DENIED {
		if _, err := os.Stat(renamePath); err == nil {
			// windows os.Rename() issue #1920
			// the destination already exists, just remove the target and continue
			return os.Remove(w.Name())
		}
	}
	return err
```

I do have to say though of the two solutions I think I'm leaning towards the first one, as it makes os.Rename behave more like linux (in that it's an atomic transaction.) Would you mind if I left this up for a second opinion? I'm more than happy to swap over to this approach, but I worry it would mask other issues outside of the scope of this ticket (as you mentioned, running two crane processes at the same time, it got me thinking that realistically that /should/ be an error!)

**Comment by slonopotamus** [comment]
Sure, I'm just passing by, and simply gave my thoughts.

**Comment by dgannon991** [comment]
@slonopotamus - thank you for taking the time, the thoughts were definitely appreciated, and it was great to have a confirm that it was an issue in other places!

**Comment by dgannon991** [comment]
Hi Guys,
Is there anything more I need to do on this one? Just aware it will auto close at some point!
Cheers,
David

File: pkg/v1/layout/write.go [hunk 1]
```diff
@@ -37,6 +39,8 @@ var layoutFile = `{
     "imageLayoutVersion": "1.0.0"
 }`
 
+var renameMutex sync.Mutex
```
Comment (imjasonh) [inline pkg/v1/layout/write.go:42]:
```suggestion
// renameMutex guards os.Rename calls in AppendImage on Windows only.
var renameMutex sync.Mutex
```

**Review by imjasonh — APPROVED** [review]

**Pull request merged**